common_components/linux_compat/freertos/freertos_linux.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * SPDX-FileCopyrightText: 2021-2022 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <unistd.h>
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include <pthread.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include "osal/osal_api.h"
+
+static uint64_t s_semaphore_data = 0;
+
+typedef enum queue_type_tag {
+    MUTEX_REC,
+    MUTEX,
+    SEMA,
+    QUEUE,
+} queue_type_t;
+
+struct generic_queue_handle {
+    queue_type_t type;
+    size_t item_size;
+    void *q;
+};
+
+static struct generic_queue_handle *create_generic_queue(queue_type_t type, uint32_t len, uint32_t item_size)
+{
+    struct generic_queue_handle *h = calloc(1, sizeof(struct generic_queue_handle));
+    h->item_size = len;
+    h->type = type;
+    switch (type) {
+    default:
+    case QUEUE:
+    case SEMA:
+        h->q = osal_queue_create();
+        break;
+
+    case MUTEX:
+    case MUTEX_REC:
+        h->q = osal_mutex_create();
+        break;
+    }
+    return h;
+}
+
+QueueHandle_t xQueueCreate(uint32_t uxQueueLength, uint32_t uxItemSize )
+{
+    return (QueueHandle_t)create_generic_queue(QUEUE, uxQueueLength, uxItemSize);
+}
+
+uint32_t xQueueSend(QueueHandle_t xQueue, const void *pvItemToQueue, TickType_t xTicksToWait)
+{
+    struct generic_queue_handle *h = xQueue;
+    return osal_queue_send(h->q, (uint8_t *)pvItemToQueue, h->item_size) ? pdTRUE : pdFAIL;
+}
+
+uint32_t xQueueSendToBack(QueueHandle_t xQueue, const void *pvItemToQueue, TickType_t xTicksToWait )
+{
+    return xQueueSend(xQueue, pvItemToQueue, xTicksToWait);
+}
+
+uint32_t xQueueReceive(QueueHandle_t xQueue, void *pvBuffer, TickType_t xTicksToWait)
+{
+    struct generic_queue_handle *h = xQueue;
+    return osal_queue_recv(h->q, (uint8_t *)pvBuffer, h->item_size, xTicksToWait) ? pdTRUE : pdFAIL;
+}
+
+BaseType_t xSemaphoreGive( QueueHandle_t xQueue)
+{
+    struct generic_queue_handle *h = xQueue;
+    if (h->type == MUTEX) {
```

Comment:
```suggestion
    if (h->type == MUTEX || h->type == MUTEX_REC) {
```

Reply:
I used this as basis for the client evaluation with heaptrack, and this was the solution for one problem that I was having, If I remember correctly the client was hanging. Maybe review the call chain? I tracked down that both were calling this function, hence the suggested change.